File: algorithms/588. Design In-Memory File System/solution.cc
class FileSystem {
    struct SimpleFile {
        map<string, SimpleFile *> subFiles;
        string name;
        string content;
        bool isFile;
        SimpleFile(string name): isFile(false) { this->name = name; }
    };
    SimpleFile * root;
public:
    FileSystem() {
        root = new SimpleFile("");
    }

    vector<string> split(const string & s, char delim) {
        vector<string> r;
        istringstream ss(s);
        string token;
        while ( getline(ss, token, delim) ) {
            if ( token.size() > 0 ) {
                r.push_back(token);
            }
        }
        return r;
    }

    vector<string> ls(string path) {
        auto cur = root;
        for ( auto & name : split(path, '/') ) {
            cur = cur->subFiles[name];
        }
        if ( cur->isFile ) {
            return {cur->name};
        } else {
            vector<string> files;
            for ( auto & p : cur->subFiles ) {
                files.push_back(p.first);
            }
            sort(files.begin(), files.end());
            return files;
        }
    }

    void mkdir(string path) {
        auto cur = root;
        for ( auto & name : split(path, '/') ) {
            if ( cur->subFiles.find(name) == cur->subFiles.end() ) {
                cur->subFiles[name] = new SimpleFile(name);
            }
            cur = cur->subFiles[name];
        }
    }

    void addContentToFile(string filePath, string content) {
        auto cur = root;
        for ( auto & name : split(filePath, '/') ) {
            if ( cur->subFiles.find(name) == cur->subFiles.end() ) {
                cur->subFiles[name] = new SimpleFile(name);
            }
            cur = cur->subFiles[name];
        }
        cur->isFile = true;
        cur->content += content;
    }

    string readContentFromFile(string filePath) {
        auto cur = root;
        for ( auto & name : split(filePath, '/') ) {
            cur = cur->subFiles[name];
        }
        return cur->content;
    }
};

/**
 * Your FileSystem object will be instantiated and called as such:
 * FileSystem obj = new FileSystem();
 * vector<string> param_1 = obj.ls(path);
 * obj.mkdir(path);
 * obj.addContentToFile(filePath,content);
 * string param_4 = obj.readContentFromFile(filePath);
 */

File: algorithms/588. Design In-Memory File System/solution.py
class FileSystem(object):

    def __init__(self):
        self.root = {}

    def ls(self, path):
        """
        :type path: str
        :rtype: List[str]
        """
        if path == '/':
            return sorted(self.root.iterkeys())
        P = path.split('/')[1:]
        d = self.root
        for p in P:
            d = d[p]
        return sorted(d.iterkeys()) if isinstance(d, dict) else [P[-1]]

    def mkdir(self, path):
        """
        :type path: str
        :rtype: void
        """
        d = self.root
        for p in path.split('/')[1:]:
            if p not in d:
                d[p] = {}
            d = d[p]

    def addContentToFile(self, filePath, content):
        """
        :type filePath: str
        :type content: str
        :rtype: void
        """
        P = filePath.split('/')[1:]
        d = self.root
        for p in P[:-1]:
            d = d[p]
        if P[-1] not in d:
            d[P[-1]] = ''
        d[P[-1]] += content

    def readContentFromFile(self, filePath):
        """
        :type filePath: str
        :rtype: str
        """
        d = self.root
        for p in filePath.split('/')[1:]:
            d = d[p]
        return d


# Your FileSystem object will be instantiated and called as such:
# obj = FileSystem()
# param_1 = obj.ls(path)
# obj.mkdir(path)
# obj.addContentToFile(filePath,content)
# param_4 = obj.readContentFromFile(filePath)
